plugins/catalog-backend/src/service/CatalogBuilder.ts

```diff
@@ -0,0 +1,523 @@
+  Validators,
+} from '@backstage/catalog-model';
+import { Config } from '@backstage/config';
+import lodash from 'lodash';
+import { Logger } from 'winston';
+import {
+  DatabaseEntitiesCatalog,
+  DatabaseLocationsCatalog,
+  EntitiesCatalog,
+  LocationsCatalog,
+} from '../catalog';
+import { DatabaseManager } from '../database';
+import {
+  AnnotateLocationEntityProcessor,
+  AzureApiReaderProcessor,
+  BitbucketApiReaderProcessor,
+  CodeOwnersProcessor,
+  EntityPolicyProcessor,
+  FileReaderProcessor,
+  GithubOrgReaderProcessor,
+  GithubReaderProcessor,
+  GitlabApiReaderProcessor,
+  GitlabReaderProcessor,
+  HigherOrderOperation,
+  HigherOrderOperations,
+  LocationProcessor,
+  LocationReaders,
+  LocationRefProcessor,
+  PlaceholderProcessor,
+  PlaceholderResolver,
+  StaticLocationProcessor,
+  UrlReaderProcessor,
+  YamlProcessor,
+} from '../ingestion';
+import { CatalogRulesEnforcer } from '../ingestion/CatalogRules';
+import { LdapOrgReaderProcessor } from '../ingestion/processors/LdapOrgReaderProcessor';
+import {
+  jsonPlaceholderResolver,
+  textPlaceholderResolver,
+  yamlPlaceholderResolver,
+} from '../ingestion/processors/PlaceholderProcessor';
+
+export type CatalogEnvironment = {
+  logger: Logger;
+  database: PluginDatabaseManager;
+  config: Config;
+  reader: UrlReader;
+};
+
+/**
+ * A builder that helps wire up all of the component parts of the catalog.
+ *
+ * The touch points where you can replace or extend behavior are as follows:
+ *
+ * - Reader processors can be added or replaced. These implement the
+ *   functionality of reading raw data from a location, in the form of an
+ *   entity definition file.
+ * - Parser processors can be added or replaced. These accept the raw data as
+ *   read by the previous processors and parse it into raw structured data
+ *   (for example, from a binary buffer containing yaml text, to a JS object
+ *   structure).
+ * - Placeholder resolvers can be replaced or added. These run on the raw
+ *   structured data between the parsing and pre-processing steps, to replace
+ *   dollar-prefixed entries with their actual values (like $file).
+ * - Pre-processors can be added or replaced. These take the raw unvalidated
+ *   data from the parser processors and can enrich or extend it before
+ *   validation. This is the place where for example codeowners data can be
+ *   injected into partial entity definitions.
+ * - Entity policies can be added or replaced. These are the first line of
+ *   validation from the output of the pre-processing step. All policies are
+ *   given the chance to inspect the entity, and all of them have to pass in
+ *   order for the entity to be considered valid from an overall point of
+ *   view.
+ * - Field format validators can be replaced. These check the format of
+ *   individual core fields such as metadata.name, such that they adhere to
+ *   certain rules.
+ * - Entity kinds can be added or replaced. These are the second line of
```

Comment:
Because policies match allOf, kinds match oneOf

Reply:
Is that because of the way things are? Thinking we could switch entity kind policies to skip entities of kinds that they don't recognize. That way it's a bit easier to layer multiple `Component` policies if needed. Again though, not sure it's the way to go wrt verification. Still provides more flexibility though right?

Reply:
The driving code won't know if returning undefined is fatal (a validate) or entirely expected (a kind)

Reply:
And the skipping part, that's how it works today